perf(transform/styled-components): optimize removal loop

overlookmotel opened the pull request:
Follow-on after #12256. Optimize the `retain` loop for removing quasis and expressions by doing bounds check only once at start, rather than on each turn of the loop.

Dunqing reviewed (approved):
Hey man, you don't let go of any instruction that can be removed.

overlookmotel committed:
perf(transform/styled-components): optimize removal loop (#12284)

Follow-on after #12256. Optimize the `retain` loop for removing quasis and expressions by doing bounds check only once at start, rather than on each turn of the loop.

overlookmotel commented:
What the hell going on with this stack? Graphite keeps refusing to merge it even though CI passes.

Dunqing commented:
> Yeah maybe I am too obsessive!

No! This way can `oxc` get faster and faster.
